MatrixKit/Controllers/MXKRoomViewController.m

```diff
@@ -3323,21 +3325,79 @@ - (void)showAttachmentInCell:(UITableViewCell*)cell
+
+                        documentInteractionController = [UIDocumentInteractionController interactionControllerWithURL:fileURL];
+                        [documentInteractionController setDelegate:self];
+                        currentSharedAttachment = selectedAttachment;
+
+                        if (![documentInteractionController presentPreviewAnimated:YES])
                         {
-                            documentInteractionController = nil;
-                            [selectedAttachment onShareEnded];
-                            currentSharedAttachment = nil;
+                            if (![documentInteractionController presentOptionsMenuFromRect:self.view.frame inView:self.view animated:YES])
+                            {
+                                documentInteractionController = nil;
+                                [selectedAttachment onShareEnded];
+                                currentSharedAttachment = nil;
+                            }
                         }
+                    };
+
+                    if (roomDataSource.mxSession.crypto
+                        && [selectedAttachment.contentInfo[@"mimetype"] isEqualToString:@"text/plain"]
+                        && [MXMegolmExportEncryption isMegolmKeyFile:fileURL])
+                    {
+                        // The file is a megolm key file
+                        // Ask the user if they wants to view the file as a classic file attachment
+                        // or open an import process
+                        [currentAlert dismiss:NO];
+
+                        __weak typeof(self) weakSelf = self;
+                        currentAlert = [[MXKAlert alloc] initWithTitle:@""
+                                                               message:[NSBundle mxk_localizedStringForKey:@"attachment_e2e_keys_file_prompt"]
+                                                                 style:MXKAlertStyleAlert];
+
+                        [currentAlert addActionWithTitle:[NSBundle mxk_localizedStringForKey:@"view"] style:MXKAlertActionStyleDefault handler:^(MXKAlert *alert)
+                         {
+                             // View file content
+                             if (weakSelf)
+                             {
+                                 typeof(self) self = weakSelf;
+                                 self->currentAlert = nil;
+
+                                 viewAttachment();
+                             }
+                         }];
+
+                        [currentAlert addActionWithTitle:[NSBundle mxk_localizedStringForKey:@"attachment_e2e_keys_import"] style:MXKAlertActionStyleDefault handler:^(MXKAlert *alert)
+                         {
+                             if (weakSelf)
+                             {
+                                 typeof(self) self = weakSelf;
+                                 self->currentAlert = nil;
+
+                                 __weak typeof(self) weakSelf2 = self;
```

Comment:
If we follow that "weak/strong self" design, we need to apply it every time we made an async thing. Objects can be released and lost in the middle on the 2nd async operation. No?